concordium-consensus/src/Concordium/MultiVersion.hs

```diff
@@ -999,48 +999,79 @@ sendCatchUpStatus genIndex = MVR $ \mvr@MultiVersionRunner{..} -> do
 -- During a protocol update, it might be wise to cache messages for
 -- the forthcoming protocol version. For the time being, we will
 -- simply rely on catch-up.
-withLatestExpectedVersion ::
+withLatestExpectedVersion' ::
     GenesisIndex ->
-    (EVersionedConfiguration gsconf finconf -> MVR gsconf finconf UpdateResult) ->
-    MVR gsconf finconf UpdateResult
-withLatestExpectedVersion gi a = do
+    (EVersionedConfiguration gsconf finconf -> MVR gsconf finconf (Skov.UpdateResult, Maybe a)) ->
+    MVR gsconf finconf (Skov.UpdateResult, Maybe a)
+withLatestExpectedVersion' gi a = do
     vvec <- liftIO . readIORef =<< asks mvVersions
     -- Length is an Int and GenesisIndex is a Word32.
     -- Assuming a 64-bit system, there is no risk of over/underflow.
     case compare (Vec.length vvec - 1) (fromIntegral gi) of
         EQ -> a (Vec.last vvec)
-        LT -> return ResultInvalidGenesisIndex
-        GT -> return ResultConsensusShutDown
+        LT -> return (Skov.ResultInvalidGenesisIndex, Nothing)
+        GT -> return (Skov.ResultConsensusShutDown, Nothing)
+
+withLatestExpectedVersion ::
+    GenesisIndex ->
+    (EVersionedConfiguration gsconf finconf -> MVR gsconf finconf Skov.UpdateResult) ->
+    MVR gsconf finconf Skov.UpdateResult
+withLatestExpectedVersion gi a = fst <$> withLatestExpectedVersion' gi (fmap (,Nothing) <$> a)
+
+-- |A continuation for executing a block that has been received
+-- and verified.
+newtype ExecuteBlock = ExecuteBlock {runBlock :: IO Skov.UpdateResult}
 
 -- |Deserialize and receive a block at a given genesis index.
-receiveBlock :: GenesisIndex -> ByteString -> MVR gsconf finconf UpdateResult
-receiveBlock gi blockBS = withLatestExpectedVersion gi $
+-- Return a continuation ('Maybe ExecuteBlock') iff. the 'Skov.UpdateResult' is 'ResultSuccess'.
+--
+-- An initial write lock is acquired for receiving the block.
+-- This is used for marking blocks as pending or even dead
+--
+-- The acquiring of the write lock for the later execution of the block is
+-- yielded within the actual continuation for doing so.
+--
+-- The continuation is expected to be invoked via 'executeBlock'.
```

Comment:
But are you sure that is the case. Because the function calls `runSkovTransaction` which both acquires and releases the lock.

Then what is returned is another computation that acquires and releases the lock again.

I don't know what was intended, but the documentation and the code do not seem to agree.

Reply:
No. The intention is what's in the code. I will clear up this documentation. 

But I am ready to discuss whether my intention is correct. 

My reasoning for acquiring the lock twice as opposed to once is that I wanted to release the lock quickest possible. But I am not sure it matters either way. Maybe you are biased towards one or another? Or @td202

Reply:
With how the node works at the moment the only benefit of releasing it is that some transactions can be injected in the time between receiveBlock terminates and executeBlock starts.

The main benefits for having one lock would be 
- make sure no state has changed in between. So easier to make sure that nothing important has changed. 
- as a result we could carry over some state from receive to execute.

The downside is that the lifetime is a bit more complex. So this might not be worthwhile.

Reply:
I presume a block awaiting its parent to be alive could also be squeezed in, same with e.g. a fin. record? And yes you are correct regarding the lifetime management of the lock but also general understanding of the flow. 

I suppose if we put everything under 1 lock then we could omit a lookup on a block (the parent of the one that is about to being added to the tree) However we use `getRecentBlockStatus´ there as opposed to ´getBlockStatus` which should be relatively cheap and not infer an overhead in the greater picture. 


But I have not tested/benchmarked the two different approaches. I found this as the most straight forward approach to handle it.

Reply:
I think blocks only arrive via that one queue. And you are blocking on that during the entire processing of the block.

Reply:
I have cleaned up the documentation.